ledger/acctupdates.go

```diff
@@ -416,6 +417,131 @@ func (au *accountUpdates) lookupKv(rnd basics.Round, key string, synchronized bo
 	}
 }
 
+func (au *accountUpdates) LookupKeysByPrefix(round basics.Round, keyPrefix string, maxKeyNum uint64) ([]string, error) {
+	return au.lookupKeysByPrefix(round, keyPrefix, maxKeyNum, true /* take lock */)
+}
+
+func (au *accountUpdates) lookupKeysByPrefix(round basics.Round, keyPrefix string, maxKeyNum uint64, synchronized bool) (resultKeys []string, err error) {
+	var results map[string]bool
+	// keep track of the number of result key with value
+	var resultCount uint64
+
+	needUnlock := false
+	if synchronized {
+		au.accountsMu.RLock()
+		needUnlock = true
+	}
+	defer func() {
+		if needUnlock {
+			au.accountsMu.RUnlock()
+		}
+		// preparation of result happens in deferring function
+		// prepare result only when err != nil
+		if err == nil {
+			resultKeys = make([]string, 0, resultCount)
+			for resKey, isValid := range results {
+				if isValid {
```

Comment:
Not a big deal, but I want readers to think of the flag as saying whether the box is present or not, not as saying whether it's "valid".
```suggestion
			for resKey, present := range results {
				if present {
```

Reply:
changed as you suggested.